xdsl/transforms/memref_stream_legalize.py

```diff
@@ -0,0 +1,174 @@
+from dataclasses import dataclass
+from typing import TypeAlias
+
+from xdsl.context import MLContext
+from xdsl.dialects import memref_stream
+from xdsl.dialects.builtin import (
+    ArrayAttr,
+    Float16Type,
+    Float32Type,
+    Float64Type,
+    IntegerAttr,
+    ModuleOp,
+    VectorType,
+    f16,
+    f32,
+)
+from xdsl.dialects.linalg import IteratorType
+from xdsl.ir import Attribute, Block, Operation
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    GreedyRewritePatternApplier,
+    PatternRewriter,
+    PatternRewriteWalker,
+    RewritePattern,
+    op_type_rewrite_pattern,
+)
+from xdsl.utils.exceptions import DiagnosticException
+from xdsl.utils.hints import isa
+
+StreamingVectorLegalizationType: TypeAlias = VectorType[
+    Float64Type | Float32Type | Float16Type
+]
+StreamingAlreadyLegalType: TypeAlias = Float64Type
+
+
+def _is_legal_vector(attr: Attribute) -> bool:
+    if isa(attr, StreamingVectorLegalizationType):
+        match attr.element_type, attr.element_count():
+            case Float64Type(), 1:
+                return True
+            case Float32Type(), 2:
+                return True
+            case Float16Type(), 4:
+                return True
+            case _:
+                return False
+    return False
+
+
+def _is_legal_attr(attr: Attribute) -> bool:
+    return _is_legal_vector(attr) or isinstance(attr, StreamingAlreadyLegalType)
+
+
+def _legalize_attr(
+    attr: Attribute,
+) -> StreamingVectorLegalizationType | StreamingAlreadyLegalType:
+    if isa(attr, StreamingVectorLegalizationType):
+        # Either already a legal vector or impossible to legalize
+        if not _is_legal_vector(attr):
+            raise DiagnosticException(f"Cannot legalize {attr} for streaming")
+        return attr
+    elif isinstance(attr, Float64Type):
+        return attr
+    elif isinstance(attr, Float32Type):
+        return VectorType(f32, (2,))
+    elif isinstance(attr, Float16Type):
+        return VectorType(f16, (4,))
+    else:
+        raise DiagnosticException(f"Cannot legalize {attr} for streaming")
+
+
+def _legalize_block(block: Block, rewriter: PatternRewriter) -> None:
+    # Start from all uses of all block arguments
+    to_be_legalized: set[Operation] = {
+        use.operation for arg in block.args for use in arg.uses
+    }
+    # Linearly scan the block and update the set of ops that we need to look into
+    # by following uses
+    for op in block.ops:
+        if op not in to_be_legalized:
+            continue
+        to_be_legalized.remove(op)
+        illegal_results: list[int] = [
+            result.index for result in op.results if not _is_legal_attr(result.type)
+        ]
+        if not illegal_results:
+            continue
+        new_op = op.create(
+            operands=op.operands,
+            result_types=[_legalize_attr(res.type) for res in op.results],
+            attributes=op.attributes,
+        )
```

Comment:
It feels like we should check for a set of operations we know this is safe to do, like `FloatingPointLikeBinaryOp` in arith.

Reply:
It looks both `isa` and `isinstance` break with the nice `FloatingPointLikeBinaryOp`, resorted to enumeration of operations known to be safe to stream-legalize.